### PR TITLE
feat(schema): support accessing Portex types in package by attr-style

### PR DESCRIPTION
**Pull request by linjiX**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2dn0caa 支持通过 "." 方法获取 package 中的类型](https://app.clickup.com/t/2dn0caa)

File: graviti/utility/attr.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    overload,
+)
+
+from tensorbay.utility import ReprMixin, ReprType
+
+_T = TypeVar("_T")
+_D = TypeVar("_D")
+
+
+class _AttrDict(Generic[_T], ReprMixin):
```
Comment (Lee-000) [inline graviti/utility/attr.py:27]: Should the order of the methods follow the standard?

File: graviti/utility/attr.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            return self._data.__getitem__(key)
+
+        return self._modules.__getitem__(prefix).__getitem__(suffix)
+
+    def _repr_head(self) -> str:
+        return AttrDict.__name__
+
+
+class AttrDict(_AttrDict[_T], Mapping[str, _T]):
+    """A dict which allows for attr-style access of values."""
```
Comment (Lee-000) [inline graviti/utility/attr.py:114]: `A dict that allows attr-style access of values.`?

File: graviti/utility/attr.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            self._getitem(key)
+        except KeyError:
+            return False
+
+        return True
+
+    def __getitem__(self, key: str) -> _T:
+        try:
+            return self._getitem(key)
+        except KeyError:
```
Comment (Lee-000) [inline graviti/utility/attr.py:59]: Would there be `AttributeError`?

File: graviti/utility/attr.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            module = _AttrDict()
+            self._modules.__setitem__(prefix, module)
+
+        module.__setitem__(suffix, value)
+
+    def __getattr__(self, key: str) -> Union["_AttrDict[_T]", _T]:
+        try:
+            return self._data.__getitem__(key)
+        except KeyError:
+            pass
```
Comment (Lee-000) [inline graviti/utility/attr.py:86]: Which one would be better, `try except` or `__contains__`?

File: graviti/utility/attr.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            return self._getitem(key)
+        except KeyError:
+            return self.__missing__(key)
+
+    def __setitem__(self, key: str, value: _T) -> None:
+        try:
+            prefix, suffix = key.split(".", 1)
+        except ValueError:
+            if key in self._modules:
+                raise KeyError("Key prefix conflict") from None
```
Comment (Lee-000) [inline graviti/utility/attr.py:67]: Setting an entire module is not allowed, but setting an item in one module is accepted?

**Review by Lee-000 — CHANGES_REQUESTED** [review]

File: graviti/utility/attr.py (hunk 3 above)
Comment (linjiX) [reply graviti/utility/attr.py:59]: `__getitem__` should raise `KeyError`

File: graviti/utility/attr.py (hunk 2 above)
Comment (linjiX) [reply graviti/utility/attr.py:114]: copyed from another python library `AttrDict`
https://github.com/bcj/AttrDict/blob/8c1883162178a124ee29144ca7abcd83cbd9d222/attrdict/mixins.py#L159

Comment (Lee-000) [reply graviti/utility/attr.py:114]: Use `that` instead of `which`?

File: graviti/utility/attr.py (hunk 4 above)
Comment (linjiX) [reply graviti/utility/attr.py:86]: `__contains__` is faster when accessing non-leaf node
`try except` is faseter when getting leaf node

File: graviti/utility/attr.py (hunk 5 above)
Comment (linjiX) [reply graviti/utility/attr.py:67]: `AttrDict` does not provide the ability to set non-leaf node manually,
`__setitem__` will treat all values input from user as leaf node even it is another `AttrDict`

**Review by Lee-000 — APPROVED** [review]

**Pull request merged**